apollo-router/src/services/subgraph_service.rs

```diff
@@ -597,104 +593,205 @@ async fn call_http(
+        .await
+        .map(|r| {
+            context.leave_active_request();
+            r
+        })?;
 
-        let body = match hyper::body::to_bytes(body)
-            .instrument(tracing::debug_span!("aggregate_response_data"))
-            .await {
-                Err(err) => {
-                    cloned_context.leave_active_request();
+    // Print out debug for the response
+    if display_headers {
+        tracing::info!(
+            http.response.headers = ?parts.headers, apollo.subgraph.name = %service_name, "Response headers from subgraph {service_name:?}"
+        );
+    }
 
-                    tracing::error!(fetch_error = format!("{err:?}").as_str());
+    if display_body {
+        if let Ok(body) = &body {
+            tracing::info!(
+                http.response.body = %String::from_utf8_lossy(body), apollo.subgraph.name = %service_name, "Raw response body from subgraph {service_name:?} received"
+            );
+        }
+    }
 
-                return Err(FetchError::SubrequestHttpError {
-                    status_code: None,
-                    service: service_name.clone(),
-                    reason: err.to_string(),
-                }.into())
+    // Parse the response.
+    // https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#processing-the-response
+    // If the response uses a non-200 status code and the media type of the response payload is application/json then the client MUST NOT rely on the body to be a well-formed GraphQL response since the source of the response may not be the server but instead some intermediary such as API gateways, proxies, firewalls, etc.
 
-                }, Ok(body) => body,
-            };
+    enum ContentType {
+        ApplicationJson,
+        ApplicationGraphqlJson,
+        Other,
+    }
+    let content_type = parts
+        .headers
+        .get(CONTENT_TYPE)
+        .and_then(|value| value.to_str().ok())
+        .map(|value| {
+            if value.contains(APPLICATION_JSON.essence_str()) {
+                ContentType::ApplicationJson
+            } else if value.contains(GRAPHQL_JSON_RESPONSE_HEADER_VALUE) {
+                ContentType::ApplicationGraphqlJson
+            } else {
+                ContentType::Other
+            }
+        })
+        .unwrap_or(ContentType::Other);
 
-            cloned_context.leave_active_request();
+    enum HttpStatus {
+        Success,
+        Error,
+    }
+    let http_status = if parts.status.is_success() {
+        HttpStatus::Success
+    } else {
+        HttpStatus::Error
+    };
 
-        Ok((parts, body))
-    }.instrument(subgraph_req_span).await?;
+    let mut graphql: graphql::Response = match (content_type, http_status, body) {
+        (ContentType::ApplicationJson, HttpStatus::Error, Ok(bytes)) => {
+            graphql::Response::from_bytes(service_name, bytes)
+                .unwrap_or_else(|_error| graphql::Response::builder().build())
```

Comment:
why not build an error response like `FetchError::SubrequestMalformedResponse` here? The status code is an error, and the content type was JSON yet we got a parse error, it looks like it warrants reporting an error

Reply:
Yeah, this was the thing highlighted in the description.
The issue is that according to the spec it's "OK" for a subgraph response to be tampered with and return non-graphql if the content_type is `ApplicationJson`.
In this situation I'm tempted to say that if the response fails to parse then the error should contain the entire response rather than the parse error.

Reply:
I've basically rewritten everything, so it needs to be reviewed from scratch.